docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,456 @@
+# RFC 017: ABCI++ Vote Extension Propagation
+
+## Changelog
+
+- 11-Apr-2022: Initial draft (@sergio-mena).
+
+## Abstract
+
+According to the
+[ABCI++ specification](https://github.com/tendermint/tendermint/blob/4743a7ad0/spec/abci%2B%2B/README.md)
+(as of 11-Apr-2022), a validator MUST provide a signed vote extension for each non-`nil` precommit vote
+of height *h* that it uses to propose a block in height *h+1*. When a validator is up to
+date, this is easy to do, but when a validator needs to catch up this is far from trivial as this data
+cannot be retrieved from the blockchain.
+
+This RFC presents and compares the different options to address this problem, which have been proposed
+in several discussions by the Tendermint Core team.
+
+## Document Structure
+
+The RFC is structured as follows. In the [Background](#background) section,
+subsections [Problem Description](#problem-description) and [Cases to Address](#cases-to-address)
+explain the problem at hand from a high level perspective, i.e., abstracting away from the current
+Tendermint implementation. In contrast, subsection
+[Current Catch-up Mechanisms](#current-catch-up-mechanisms) delves into the details of the current
+Tendermint code.
+
+In the [Discussion](#discussion) section, subsection [Solutions Proposed](#solutions-proposed) is also
+worded abstracting away from implementation details, whilst subsections
+[Feasibility of the Proposed Solutions](#feasibility-of-the-proposed-solutions) and
+[Current Limitations and Possible Implementations](#current-limitations-and-possible-implementations)
+analize the viability of one of the proposed solutions in the context of Tendermint's architecture
+based on reactors. Finally, [Formalization Work](#formalization-work) brifely discusses the work
+still needed demonstrate the correctness of the chosen solution.
+
+## Background
+
+This document assumes that all validators have equal voting power for the sake of simplicity. This is done
+without loss of generality.
+
+### Basic Definitions
+
+TODO: Precommit votes, precommit messages
+
+TODO: Re-read "commits do not contain nil votes", catchup messages
+
+TODO: Mention somewhere that statesync is not really relevant for this problem... only for the first block after statesync'ing.
+
+### Problem Description
+
+In the version of [ABCI](https://github.com/tendermint/spec/blob/4fb99af/spec/abci/README.md) present up to
+v0.35, for any height *h*, a validator *v* MUST have the decided block *b* and a *commit* for height *h* in
+order to decide at height *h*, and propose at height *h+1* in the rounds of *h+1* where *v* is a proposer.
+This *commit* consists of more than *2n/3* precommit votes voting for *b*, where *n* is the size of the
+validator set at height *h*.
```

Comment:
Actually, what I did is that, at the beginning I assumed "all validators have the same voting power", so that I could use simpler expressions for "two thirds" all over the text. Did this because the particular voting power of each validator doesn't really matter for the discussions in this document.